File: ATVD02_DAWKINS_WEASEL/dawkins_weasel.py
#Implementação do Richard Dawkins's weasel program
#Feito por Karla Félix para a disciplina de Paradigmas de Linguagem de Programação
#Turma ESTCMP005 2020-2

import random
import string
import sys

CONST_TARGET = list("METHINKS IT IS LIKE A WEASEL")  # String alvo
TARGET_LEN = len(CONST_TARGET)
SIZE = range(100) #Numero de vezes que a string vai ser gerada
PROB = 0.05 #Porcentagem de probabilidade de mutação
LETTERS = string.ascii_uppercase + ' '

highestScoreStr = []


def generateRandomString():
    evoStr = []
    for _ in range(TARGET_LEN):
        evoStr.append(random.choice(LETTERS))
    return evoStr

def mutate(evoStr):
    # gera um novo caracter para cara um dos caracteres de cada uma das 100 cópias
    return [(random.choice(LETTERS) if random.random() < PROB else evoChar) for evoChar in evoStr]

def comparison(evoStr):
    score = 0
    if len(evoStr) != 0:
        for i in range (TARGET_LEN):
            if CONST_TARGET[i] == evoStr[i]:
                score += 1
                
    return score

def reproduce(evoStr):
    global highestScoreStr
    for _ in SIZE:
        evoStrCopy = mutate(evoStr)
        score = comparison(evoStrCopy)
        if score > comparison(highestScoreStr):
            highestScoreStr = evoStrCopy
        
    return highestScoreStr, comparison(highestScoreStr)

def main():
    tries, score = 0, 0
    evolutionString = generateRandomString()
    while evolutionString != CONST_TARGET:
        tries += 1
        evolutionString, score = reproduce(evolutionString)
        print(f"Tentativa {tries}: {''.join(highestScoreStr)} -- Score: {score}")
    else:
        print(f"O algortimo de Weasel levou {tries} gerações para conseguir o reproduzir a frase!")

main()

File: ATVD02_DAWKINS_WEASEL/dawkins_weasel.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

const char target[] = "METHINKS IT IS LIKE A WEASEL";//target 
const char letters[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZ ";//letras do alfabeto e espaço em branco

#define TARGET_SIZE strlen(target)
#define SIZE 100
#define MUTATION_PROB 5

/* retorna um inteiro aleatorio */
int irand(int n)
{
    int r, rand_max = RAND_MAX - (RAND_MAX % n);
    while ((r = rand()) >= rand_max);
    return r / (rand_max / n);
}

char *reproduce(char *evoStr)
{
    char *evoStrCopy = malloc(TARGET_SIZE * sizeof(char));
    char *highestScore = (char *)malloc(TARGET_SIZE * sizeof(char));
    int score = 0;
    int cont = 0;

    for (int i = 0; i < SIZE; i++)
    {
        strncpy(evoStrCopy, evoStr, TARGET_SIZE);
        //mutate
        for (int i = 0; i < TARGET_SIZE; i++)
        {
            if (irand(100) <= MUTATION_PROB)
            {
                evoStrCopy[i] = letters[irand(TARGET_SIZE)];
            }
        }
        score = comparison(evoStrCopy);
        if (score > comparison(highestScore))
        {
            strncpy(highestScore, evoStrCopy, TARGET_SIZE);
        }
    }

    return highestScore;
}

int comparison(char *evoStr)
{
    int score = 0;
    for (int i = 0; i < TARGET_SIZE; i++)
    {
        if (target[i] == evoStr[i])
        {
            score++;
        }
    }
    return score;
}

int main()
{
    int tries = 0;
    char *evolutionString = (char *)malloc(TARGET_SIZE * sizeof(char));

    /*Random String*/
    for (int i = 0; target[i]; i++)
    {
        evolutionString[i] = letters[irand(TARGET_SIZE)];
    }
    int cont = 0;
    do
    {
        evolutionString = reproduce(evolutionString);
        tries++;
        printf("Tentativa %d : %s score: %d \n", tries, evolutionString, comparison(evolutionString));
    } while (comparison(evolutionString) != TARGET_SIZE);
    printf("O Algortimo de Weasel levou %d tentativas para reproduzir a frase!");
    return 0;
}
